src/sdsc/xslt.py

```diff
@@ -0,0 +1,396 @@
+    """
+
+    if isinstance(word, list):
+        if start:
+            word = [removepunctuation(word[0], start=True)] + word[1:]
+        if end:
+            word = word[:-1] + [removepunctuation(word[-1], end=True)]
+        return word
+
+    startpunctuation = '([{"\'¡¿'
+    endpunctuation = ')]}/\\"\',:;!?.‥…‼‽⁇⁈⁉'
+
+    if start:
+        word = word.lstrip(startpunctuation)
+    if end:
+        word = word.rstrip(endpunctuation)
+    return word
+
+
+def canBeDupe(word):
+    """TAGIGNORE: character sequences that should be ignored by the duplicate
+    words check
+    Ignored contents from tags can take either of the two forms below:
+       ##@lowercase-1##
+       ##@lowercase##
+    The second form is counted as one token, the first one is counted as as many
+    tokens as the number given after the dash."""
+
+    numberignore = re_compile(r'[[{(\'"\s]*[-+]?[0-9]+(?:[.,][0-9]+)*[]})\'";:.\s]*')
+
+    return len(word) > 0 and not numberignore.match(word) and not findtagreplacement(word)[0]
+
+
+def isDupe(tokens, pos):
+    """Returns how many tokens at pos are duplicated
+       tokens = ["this", "is", "is", "a", "test"]
+       pos = 2
+       return 1"""
+    # FIXME: Find a clever way to be able to check for variants of the same
+    # word, such as: a/an, singular/plural, verb tenses. It should ideally
+    # not be hardcoded here.
+    maxlen = min(3, pos, len(tokens) - pos)
+    tokens = tokens[:]
+    for l in range(1, maxlen + 1):
+        if not canBeDupe(removepunctuation(tokens[pos + l - 1], start=True, end=True)):
+            return 0
+
+        if removepunctuation(tokens[pos - l:pos], start=True) == removepunctuation(tokens[pos:pos + l], end=True):
+            return l
+
+    return 0
+
+
+def highlight(tokens, highlightstart, highlightend):
+    """TODO@sknorr
+
+    :param tokens:
+
+    Example:
+       tokens = ["highlight", "these", "two", "words"]
+       highlightstart = 1
+       highlightend = 2
+       return "highlight <highlight>these two</highlight> words"
+       tokens can be a string as well, it will be tokenized automatically."""
+
+    tokens = tokens[:] # toms: why?
+    if isinstance(tokens, str):
+        return highlight(tokenizer(tokens), highlightstart, highlightend)
+
+    if highlightstart >= len(tokens) or highlightend < highlightstart:
+        return ""  # Nothing to highlight
+
+    highlightend = min(highlightend, len(tokens) - 1)
+
+    tokens[highlightstart] = "<highlight>" + tokens[highlightstart]
+    tokens[highlightend] = tokens[highlightend] + "</highlight>"
+
+    return " ".join(tokens)
+
+
+def xmlescape(text):
+    """TODO@sknorr
+
+    :param str text:
+    """
+    escapetable = {
+        "&": "&amp;",
+        '"': "&quot;",
+        "'": "&apos;",
+        ">": "&gt;",
+        "<": "&lt;",
+        }
+    return "".join(escapetable.get(c, c) for c in text)
+
+
+def dupecheckmessage(context, quote, duplicate, contextid, basefile):
+    """TODO@sknorr"""
```

Comment:
Fixed in commit 6140347

I guess, the explanation of the parameters can also be improved. I have some ideas, but probably it's better if you fill the gaps.